cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/DataPipelineConnectionTest.java

```diff
@@ -686,6 +687,78 @@ public PreviewStatus.Status call() throws Exception {
     deleteDatasetInstance(NamespaceId.DEFAULT.dataset(sinkTableName));
   }
 
+  @Test
+  public void testConnectionsWithArgumentSetterAction() throws Exception {
+    testConnectionsWithArgumentSetterAction(Engine.MAPREDUCE);
+    testConnectionsWithArgumentSetterAction(Engine.SPARK);
+  }
+
+  private void testConnectionsWithArgumentSetterAction(Engine engine) throws Exception {
+    String sourceConnName = "sourceConn" + engine;
+    String sinkConnName = "sinkConn" + engine;
+    String srcTableName = "src" + engine;
+    String sinkTableName = "sink" + engine;
+    String actionTableName = "action" + engine;
+
+    Schema schema = Schema.recordOf("x", Schema.Field.of("name", Schema.of(Schema.Type.STRING)));
+
+    // add secure macro to verify json value of secure data does not fail the macro evaluation in connections
+    String schemaJson = schema.toString();
+    getSecureStoreManager().put(NamespaceId.DEFAULT.getNamespace(), "json", schemaJson, "", Collections.emptyMap());
+
+    addConnection(
+        sourceConnName, new ConnectionCreationRequest(
+            "", new PluginInfo("test", "dummy", null, ImmutableMap.of("tableName", srcTableName,
+            "schema", "${secure(json)}"),
+            new ArtifactSelectorConfig())));
+
+    //row1column1 is the macro key for sink table name which should be set by the action plugin
+    addConnection(
+        sinkConnName, new ConnectionCreationRequest(
+            "", new PluginInfo("test", "dummy", null, Collections.singletonMap("tableName", "${row1column1}"),
+            new ArtifactSelectorConfig())));
+
+    ETLBatchConfig config = ETLBatchConfig.builder()
+        // 'row1column1' is configured in runtime arg to 'dummy', but action will set an argument that will make it 'sinkTableName'
+        .addStage(new ETLStage("action1", MockAction.getPlugin(actionTableName, "row1", "column1",
+            String.format("%s", sinkTableName))))
+        .addStage(new ETLStage("source", MockSource.getPluginUsingConnection(sourceConnName)))
+        .addStage(new ETLStage("sink", MockSink.getPluginUsingConnection(sinkConnName)))
+        .addConnection("action1", "source")
+        .addConnection("source", "sink")
+        .build();
+    // runtime arguments
+    Map<String, String> runtimeArguments = ImmutableMap.<String, String>builder().put("srcTable", srcTableName)
+        .put("row1column1", "dummy")
+        .put("system.skip.normal.macro.evaluation", "true")
```

Comment:
It is better to first not set this argument, and verify the pipeline failed, and then set this to true and check the pipeline passed.

Reply:
ack